src/hotspot/share/opto/escape.cpp

```diff
@@ -4436,6 +4452,24 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
       _compile->get_alias_index(tinst->add_offset(oopDesc::mark_offset_in_bytes()));
       _compile->get_alias_index(tinst->add_offset(oopDesc::klass_offset_in_bytes()));
       if (alloc->is_Allocate() && (t->isa_instptr() || t->isa_aryptr())) {
+        // Update adr type captured by NarrowMem projections to new type
+        InitializeNode* init = alloc->as_Allocate()->initialization();
+        assert(init != nullptr, "can't find Initialization node for this Allocate node");
+        for (DUIterator_Fast imax, i = init->fast_outs(imax); i < imax; i++) {
+          ProjNode* proj = init->fast_out(i)->as_Proj();
+          if (proj->Opcode() == Op_NarrowMemProj) {
+            const TypePtr* adr_type = proj->adr_type();
+            const TypePtr* new_adr_type = tinst->add_offset(adr_type->offset());
+            if (adr_type != new_adr_type) {
+              uint alias_ix = _compile->get_alias_index(new_adr_type);
+              assert(_compile->get_general_index(alias_ix) == _compile->get_alias_index(adr_type), "new adr type should be narrowed down from existing adr type");
+              igvn->hash_delete(proj);
+              ((NarrowMemProjNode*)proj)->set_adr_type(new_adr_type);
+              igvn->hash_insert(proj);
+              record_for_optimizer(proj);
```

Comment:
That seems to be the only reason why we need the `_adr_type` non constant. How bad would it be if we just created a new node? Hmm maybe not worth it. Just wanted to know if you had considered it?